database/meta.go

```diff
@@ -230,6 +271,46 @@ func (m *MetaDB) GetUnAnalyzedDatabases() []string {
 	return results
 }
 
+// GetAnalyzeReadyDatabases builds a list of database names which are ready to be analyzed
+func (m *MetaDB) GetAnalyzeReadyDatabases() []string {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	ssn := m.dbHandle.Copy()
+	defer ssn.Close()
+
+	var results []string
+	var cur DBMetaInfo
+
+	//migrate old datasets without the import_finished field
+	iter := ssn.DB(m.config.S.Bro.MetaDB).C(m.config.T.Meta.DatabasesTable).Find(
+		bson.M{
+			"import_finished": bson.M{"$exists": false},
+		},
+	).Iter()
+	for iter.Next(&cur) {
+		m.lock.Unlock()
+		err := m.MarkDBImported(cur.Name, true)
+		m.lock.Lock()
+		if err != nil {
+			m.log.WithFields(log.Fields{
+				"err": err,
+			}).Error("Could not add import_finished field to old databases")
+		}
+	}
+	//end migratoin
+
+	iter = ssn.DB(m.config.S.Bro.MetaDB).C(m.config.T.Meta.DatabasesTable).Find(
+		bson.M{
+			"analyzed":        false,
+			"import_finished": true,
```

Comment:
Instead of doing the migration code in lines 284-300 above, the approach I took for a similar scenario in the agent was to change the query here to be `"import_finished": bson.M{"$ne": false},` This is equivalent to the migration above as it assumes that all non-existent keys are true. This method requires less code, and my reasoning essentially boiled down to that this is NoSQL and there are no schemas enforced.  Migrations are a necessity for SQL databases because the schema is rigid, but I didn't have a good reason to impose SQL patterns on a NoSQL database. 

I tried to find anything in the Mongo documentation that would indicate which method is idiomatic or preferred for one reason or another. I didn't come up with much, but I thought this was a great discussion on the subject.
https://stackoverflow.com/questions/49330096/schema-migration-scripts-in-nosql-databases

Bottom line for me is that I prefer the "incremental upgrade" because 1) it requires less code, 2) it's less complex (though the resulting query is a little mind bending) and 3) it doesn't require touching every object.  While #3 is not all that important here because the number of objects will be low, I thought it would be a good precedent to set for handling changes in one of the high-volume collections.  I could be swayed by a good argument in the other direction because there are a couple downsides to this approach as well (namely, you have to keep the code there indefinitely and it could become convoluted if there are more instances where this is needed since you can't ever expect the schema to be a certain way).

However, if there's a better case to be made for writing a migration script then I think it may make more sense to break this (lines 284-300) out into its own file/function.  If we add more migrations in the future it would be easier to keep track of them if various snippets weren't embedded across the codebase.

Thoughts?

Reply:
I like the approach you suggest. However, I do not believe that will work well here. 

The issue is that there is a field on the returned structs `ImportFinished`. If the field doesn't exist in the database, this field will default to false. So, the method here could return structs with the `ImportFinished` field set to false. In order to fix this, we could set `ImportFinished` to true on all of the returned structs, but we'd have to apply this fix to the methods `GetAnalyzedDatabases` and `GetUnanalyzedDatabases` as well. Unfortunately, the fix cannot be directly applied to these other functions since we won't know if the field was unset or missing. 

I agree that doing the migration in the code here is gross. Would it be better if I came up with a migration layer in the install script?